src/Pirouette/Transformations/Prenex.hs

```diff
@@ -0,0 +1,163 @@
+
+-- | Shift all the bound type variables,
+-- starting from 'from' upwards.
+shiftTyBy1 :: Integer -> Type lang -> Type lang
+shiftTyBy1 from = go
+  where
+    oneMore = shiftTyBy1 (from + 1)
+
+    go (SystemF.TyFun ty1 ty2) 
+      = SystemF.TyFun (go ty1) (go ty2)
+    go (SystemF.TyLam ann ki rest)
+      = SystemF.TyLam ann ki (oneMore rest)
+    go (SystemF.TyAll ann ki rest)
+      = SystemF.TyAll ann ki (oneMore rest)
+    go (SystemF.TyApp hd args)
+      = let shiftedArgs = map go args
+        in case hd of
+             SystemF.Bound ann i | i >= from
+               -> SystemF.TyApp (SystemF.Bound ann (i + 1)) shiftedArgs
+             _ -> SystemF.TyApp hd shiftedArgs
+
+-- | Update the body of a definition
+-- with the prenex-ed types from 'newDecls'.
+prenexDefinitionBody ::
+  Decls lang ->
+  Definition lang ->
+  Definition lang
+prenexDefinitionBody newDecls = onlyOnFunDef $ \funDef ->
+  funDef { funBody = prenexExpr newDecls (funBody funDef) } 
+
+prenexExpr ::
+  forall lang.
+  Decls lang ->
+  Term lang ->
+  Term lang
+prenexExpr newDecls = goTerm
+  where 
+    goTerm :: Term lang -> Term lang
+    goTerm (SystemF.Lam ann ty body) = SystemF.Lam ann ty (goTerm body)
+    goTerm (SystemF.Abs ann ki body) = SystemF.Abs ann ki (goTerm body)
+    goTerm (SystemF.App hd args) =
+      let prenexArgs = map goArg args
+      in case hd of
+           -- we only change the applications of known names
+           SystemF.Free (TermSig name)
+             | Just (DFunDef FunDef { funTy }) <- Map.lookup name newDecls
+             -> SystemF.App hd (zipArgs prenexArgs funTy)
+           _other -> SystemF.App hd prenexArgs
+
+    goArg (SystemF.TermArg e) = SystemF.TermArg (goTerm e)
+    goArg (SystemF.TyArg ty) = SystemF.TyArg ty -- type arguments don't change
+
+    -- no more arguments
+    zipArgs [] _ = []
+    zipArgs args (SystemF.TyAll _ _ restOfTy) = case firstTyArg args of
+      Nothing -> args -- none was found, just use arg as is
+      Just (ty, moreArgs) -> ty : zipArgs moreArgs restOfTy
+    zipArgs args (SystemF.TyFun _ restOfTy) = case firstTermArg args of
+      Nothing -> args -- none was found, just use arg as is
+      Just (ty, moreArgs) -> ty : zipArgs moreArgs restOfTy
+    zipArgs args _ = args
+
+    firstTyArg [] = Nothing
+    firstTyArg (ty@SystemF.TyArg {} : rest) = Just (ty, rest)
+    firstTyArg (other : rest) = do
+      (ty, args) <- firstTyArg rest
+      pure (ty, other : args)
+
+    firstTermArg [] = Nothing
+    firstTermArg (ty@SystemF.TermArg {} : rest) = Just (ty, rest)
+    firstTermArg (other : rest) = do
+      (ty, args) <- firstTermArg rest
+      pure (ty, other : args)
+
+-- | Apply a function only to 'FunDef's
+-- and leave the rest of definitions unchanged.
+onlyOnFunDef ::
+  (FunDef lang -> FunDef lang) ->
+  Definition lang -> Definition lang
+onlyOnFunDef f = \case
+  DFunDef funDef -> DFunDef (f funDef)
+  other          -> other
```

Comment:
This could make its way into `Pirouette.Term.Syntax.Base`; Also, given the number of such _onFunDef_, _onDataDef_, etc.. functions, might make sense to have an issue to bring in something like `microlens` at some point.

Reply:
Done in the next commit. I'm depending on `optics-core`, which is as small as `microlens`, but with better error messages.